File: Chapter 7 - Lecture/p16.m
%P16  Poisson equation on [-1,1]x[-1,1] 
%     with  u = 0 on boundary

% Set up grids and tensor product Laplacian and solve for u:
N = 32;
[D, x] = cheb(N); 
y = x;
[xx,yy] = meshgrid(x(2:N),y(2:N));
% Strecth 2D grids to 1D vectors.
xx = xx(:);
yy = yy(:);             
f = 10*sin(8*xx.*(yy-1));
D2 = D^2;
D2 = D2(2:N,2:N);
I = eye(N-1);
% L = sparse(kron(I,D2) + kron(D2,I));
L = kron(I,D2) + kron(D2,I);
figure(1), clf, spy(L), drawnow
tic, u = L\f; toc                   % solve problem and watch the clock

% Reshape long 1D results onto 2D grid:
uu = zeros(N+1,N+1); uu(2:N,2:N) = reshape(u,N-1,N-1);
[xx,yy]= meshgrid(x,y);
value = uu(N/4+1,N/4+1);

% Interpolate to finer grid annd plot:
[xxx, yyy] = meshgrid(-1:0.04:1,-1:0.04:1);
uuu = interp2(xx,yy,uu,xxx,yyy,'spline');
figure, clf, mesh(xxx,yyy,uuu), colormap([0 0 0]);
xlabel x, ylabel y, zlabel u
text(0.4,-0.3,-0.3, sprintf('u(2^{-1/2}, 2^{-1/2}) = %14.11f', value));

File: p16.py
# -*- coding: utf-8 -*-
"""
Poisson eq. on [-1,1]x[-1,1] with u = 0 on boundary

@author: nguyenl
"""

import numpy as np
import matplotlib.pyplot as plt
from chebPy import cheb
from scipy.interpolate import interp2d
# from mpl_toolkits.mplot3d import Axes3D

plt.close('all')

# Set up grids and tensor product Laplacian and solve for

N = 24
D, x = cheb(N)
y = x
xx, yy = np.meshgrid(x[1:N], y[1:N])
# xx = np.reshape(xx,(N-1)**2)
xx = xx.ravel()
yy = yy.ravel()
f = 10 * np.sin(8*xx * (yy - 1))

D2 = np.dot(D, D)
D2 = D2[1:N, 1:N]
IMatrix = np.eye(N-1)
L = np.kron(IMatrix, D2) + np.kron(D2, IMatrix)
fig = plt.figure(figsize=(10,8), dpi=80)
plt.subplot(1,2,1)
plt.spy(L)

u = np.linalg.solve(L, f)
# Reshape long 1D results onto 2D grid
uu = np.zeros((N+1, N+1))
uu[1:N, 1:N] = np.reshape(u, newshape=(N-1,N-1))
value = uu[N//4, N//4]

# Interpolate to finer grid and plot
xxx = np.linspace(-1, 1, 41)
f = interp2d(x, y, uu, kind='cubic')
uuu = f(xxx, xxx)
X, Y = np.meshgrid(xxx, xxx)
# fig = plt.figure(figsize=(8, 8), dpi=80)
ax = fig.add_subplot(1,2,2, projection='3d')
ax.plot_surface(X, Y, uuu, rstride=1, cstride=1, cmap=plt.cm.jet)

plt.title(r"$u(2^{-1/2},2^{-1/2})$="+str(value))